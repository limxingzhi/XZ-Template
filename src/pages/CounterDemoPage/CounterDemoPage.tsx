import React from 'react';
import {
  Badge,
  Alert,
} from 'antd';
import { useSelector } from 'react-redux';

import {
  Counter,
} from '../../components';

import {
  store,
  counterSlice,
  appState,
} from '../../redux/reduxCounter';

const CounterDemoPage: React.FC = () => {
  const updateReduxState = (value: number = 0, reset:boolean = false) => {
    if (reset)
      store.dispatch(counterSlice.actions.reset());
    else
      store.dispatch(counterSlice.actions.increment({ value: value }));
  }
  return (<div>
    <Badge count={useSelector((state: appState) => state.count)}>
      <Alert message="Redux is pre-configured with redux-toolkit" type="info"></Alert>
    </Badge>
    <br />
    <br />
    <Counter countChangeFunction={updateReduxState} />
  </div>);
}

export default CounterDemoPage;
