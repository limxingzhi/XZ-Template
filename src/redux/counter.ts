import {
  createSlice,
  configureStore
} from '@reduxjs/toolkit';

const counterSlice = createSlice({
  name: 'counter',
  initialState: 0,
  reducers: {
    increment: (state, action) => {
      const value =  action.payload.value ?? 1;
      state += value;
      return state;
    },
    decrement : (state, action) => {
      const value =  action.payload.value ?? 1;
      state -= value;
      return state;
    },
    reset: state => 0,
  }
});

const store = configureStore({
  reducer: counterSlice.reducer,
});

export {
  store,
  counterSlice,
};
