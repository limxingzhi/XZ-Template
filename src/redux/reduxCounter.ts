import {
  createSlice,
  configureStore
} from '@reduxjs/toolkit';

interface appState {
  count: number
};

const defaultState = {
  count: 0,
};

const counterSlice = createSlice({
  name: 'counter',
  initialState: defaultState,
  reducers: {
    increment: (state: appState, action) => {
      state.count += action.payload.value ?? 0;
      return state;
    },
    reset: (state: appState) => {
      state.count = 0;
      return state;
    },
  }
});

const store = configureStore({
  reducer: counterSlice.reducer,
});

export {
  store,
  counterSlice,
  defaultState,
};

export type {
  appState,
}
