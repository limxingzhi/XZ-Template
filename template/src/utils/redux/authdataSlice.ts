import { createSlice } from "@reduxjs/toolkit";

interface IAuthdata {
  userAuthenticated: boolean;
}

const defaultAuthdataState: IAuthdata = {
  userAuthenticated: false,
};

const authdataSlice = createSlice({
  name: "authdata",
  initialState: defaultAuthdataState,
  reducers: {
    changeAuthStatus: (state: IAuthdata, action): IAuthdata => {
      return {
        ...state,
        userAuthenticated: action.payload,
      };
    },
  },
});

export { authdataSlice, defaultAuthdataState };

export type { IAuthdata };
