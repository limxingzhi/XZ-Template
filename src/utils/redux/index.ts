import { configureStore, combineReducers } from "@reduxjs/toolkit";
import { appconfigSlice, IAppconfig } from "./appconfigSlice";
import { authdataSlice, IAuthdata } from "./authdataSlice";

interface IGlobalState {
  appconfig: IAppconfig,
  authdata: IAuthdata,
}

const store = configureStore({
  reducer: combineReducers({
    appconfig: appconfigSlice.reducer,
    authdata: authdataSlice.reducer,
  }),
  devTools: true,
});

export { store, appconfigSlice };
export type { IGlobalState };
