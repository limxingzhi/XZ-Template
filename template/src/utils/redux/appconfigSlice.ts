import { createSlice } from "@reduxjs/toolkit";
import { changeTitle } from 'utils/utils';

interface IAppconfig {
  currentPage: number;
}

interface IPageDataAction {
  type: string,
  payload: {
    key: number;
    title: string;
  }
}

const defaultAppconfigState: IAppconfig = {
  currentPage: 0,
};

const appconfigSlice = createSlice({
  name: "appconfig",
  initialState: defaultAppconfigState,
  reducers: {
    updatePage: (state: IAppconfig, action: IPageDataAction): IAppconfig => {
      changeTitle(action.payload.title);
      return {
        ...state,
        currentPage: action.payload.key,
      };
    },
  },
});

export { appconfigSlice, defaultAppconfigState };

export type { IAppconfig };

