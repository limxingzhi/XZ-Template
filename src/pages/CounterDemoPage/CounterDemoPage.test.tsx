// useful reads
// https://reactjs.org/docs/testing-recipes.html
// what does act() do https://github.com/mrdulin/react-act-examples/blob/master/sync.md

import React from 'react';
import { unmountComponentAtNode, render } from "react-dom";
import TestRenderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/react";
import { counterSlice, store } from '../../redux/reduxCounter';
import { Badge } from 'antd';
import pretty from "pretty";
import CounterDemoPage from './CounterDemoPage';

let container : HTMLElement;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);

  // reset the count for the state
  store.dispatch(counterSlice.actions.reset());
});

describe("CounterDemoPage", ()=> {
  test('default render matches snapshot', () => {
    render(<Provider store={store}>
      <CounterDemoPage/>
    </Provider>, container);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  test('clicking increments updates store correctly', async () => {
    TestRenderer.create(<Provider store={store}>
      <CounterDemoPage/>
    </Provider>);
    const incrementButton = document.querySelector("[data-testid=increment-click]");

    // making sure initial state is 0
    expect(store.getState().count).toBe(0);

    // clicks increment button
    await waitFor(() => {
      incrementButton?.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    // waits for half a second for store to be updated before checking the new value
    setTimeout(()=> {
      expect(store.getState().count).toBe(1);
    }, 500);
  });

  test('clicking increments updates badge correctly', async () => {
    const testRenderer = TestRenderer.create(<Provider store={store}>
      <CounterDemoPage/>
    </Provider>);
    const testInstance = testRenderer.root;
    const incrementButton = document.querySelector("[data-testid=increment-click]");

    // making sure badge count is 0
    expect(testInstance.findByType(Badge).props.count).toBe(0);

    // clicks increment button
    await waitFor(() => {
      incrementButton?.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    // waits for half a second for store to be updated before checking the new value
    setTimeout(()=> {
      expect(testInstance.findByType(Badge).props.count).toBe(1);
    }, 500);
  });
})

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
});
