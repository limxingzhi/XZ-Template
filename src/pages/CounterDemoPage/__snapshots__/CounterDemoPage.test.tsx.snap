// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterDemoPage default render matches snapshot 1`] = `
"<div><span class=\\"ant-badge\\"><div data-show=\\"true\\" class=\\"ant-alert ant-alert-info ant-alert-no-icon\\" role=\\"alert\\"><div class=\\"ant-alert-content\\"><div class=\\"ant-alert-message\\">Redux is pre-configured with redux-toolkit</div><div class=\\"ant-alert-description\\"></div></div></div></span><br><br>
  <div style=\\"width: 200px;\\" class=\\"ant-card ant-card-bordered ant-card-hoverable counter__card\\">
    <div class=\\"ant-card-head\\">
      <div class=\\"ant-card-head-wrapper\\">
        <div class=\\"ant-card-head-title\\">Redux Counter</div>
      </div>
    </div>
    <div class=\\"ant-card-body\\">
      <div class=\\"ant-space ant-space-horizontal ant-space-align-center\\">
        <div class=\\"ant-space-item\\" style=\\"margin-right: 24px;\\">
          <div class=\\"ant-space ant-space-horizontal ant-space-align-center\\">
            <div class=\\"ant-space-item\\" style=\\"margin-right: 8px;\\"><button data-testid=\\"increment-click\\" type=\\"button\\" class=\\"ant-btn ant-btn-circle ant-btn-sm\\"><span role=\\"img\\" aria-label=\\"caret-up\\" class=\\"anticon anticon-caret-up\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-up\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z\\"></path></svg></span></button></div>
            <div class=\\"ant-space-item\\"><button data-testid=\\"decrement-click\\" type=\\"button\\" class=\\"ant-btn ant-btn-circle ant-btn-sm\\"><span role=\\"img\\" aria-label=\\"caret-down\\" class=\\"anticon anticon-caret-down\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z\\"></path></svg></span></button></div>
          </div>
        </div>
        <div class=\\"ant-space-item\\"><button type=\\"button\\" class=\\"ant-btn ant-btn-sm\\"><span>Reset</span></button></div>
      </div>
    </div>
  </div>
</div>"
`;
